### PR TITLE
test: Add MCP Schemas Coverage Regression Test

### PR DESCRIPTION
**Pull request by waqasshaukat**
**Why needed this test?**
After AI SDK V5 bump, it is mandatory to provide the schemas of all the tools to LLM. Without them it throws input validation error. Moreover, missing to give some tool schema(s) to it, prohibit to work tools corresponds to them. This regression test has been written to tackle these scenarios.

### CODE REVIEW

File: test/mcp-schema-validation.test.ts [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+    // Return cached result if available
+    if (mcpClientCache[serverKey]) {
+        return mcpClientCache[serverKey].tools;
+    }
+
+    const serverConfig = MCP_SERVERS[serverKey as keyof typeof MCP_SERVERS];
+
+    try {
+        console.log(`\n🔌 Connecting to ${serverConfig.name}...`);
+        console.log(`   URL: ${serverConfig.url}`);
```
Comment (mujahidkay) [inline test/mcp-schema-validation.test.ts]: Let's please be vigilant in removing pointless console logs generated by AI.

**Review by mujahidkay — COMMENTED** [review]
I don't think the PR is serving its purpose. We are nowhere actually checking if the schemas of corresponding tools match. Asserting the schemas to be of type `object` and `toBeDefined` is not sufficient.

I think everything can be reconciled into one parameterized test. For each parameter, one of the three MCP servers, we need to:
1. check num of tools received from MCP server equals the number of tools in schema file
2. check if each tool's schema (from remote MCP server) matches that in the schema file

However, since (2) isn't currently testable let's just go with (1) for now until https://github.com/agoric-labs/ymax-mcp-server/pull/11 merges.

File: test/mcp-schema-coverage.test.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+describe('MCP Server Schema Coverage', () => {
+    afterAll(async () => {
+        await cleanupMcpClients(mcpClientCache, MCP_SERVERS);
+    });
+
+    Object.entries(MCP_SERVERS).forEach(([serverKey, serverConfig]) => {
+        describe(serverConfig.name, () => {
+            it('should have matching schemas for all MCP server tools', { timeout: TEST_TIMEOUT }, async () => {
+                const serverTools = await fetchMcpServerTools(serverKey, serverConfig, mcpClientCache);
```
Comment (Muneeb147) [inline test/mcp-schema-coverage.test.ts]: @waqasshaukat 

What's the usage `mcpClientCache`?
Each server's tool is being fetched only once (per test run `yarn test:mcp`).
If I run yarn test:mcp again, then `mcpClientCache` will be empty. So where is caching helping us? I don't see it's benefit whether in "intra" or "inter" test runs?

OR are you considering any future benefit from cache? Like when we have more tests running per server under same suite?

File: test/mcp-schema-coverage.test.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+const TEST_TIMEOUT = 60_000;
+
+describe('MCP Server Schema Coverage', () => {
+    afterAll(async () => {
+        await cleanupMcpClients(mcpClientCache, MCP_SERVERS);
+    });
+
+    Object.entries(MCP_SERVERS).forEach(([serverKey, serverConfig]) => {
+        describe(serverConfig.name, () => {
+            it('should have matching schemas for all MCP server tools', { timeout: TEST_TIMEOUT }, async () => {
```
Comment (Muneeb147) [inline test/mcp-schema-coverage.test.ts]: I think the test is only checking presence of schema against the tools.
Can we also verify that the schema/fields should actually match? Schema presence don't just make sure that we are covering the breaking changes.

Comment (Muneeb147) [reply test/mcp-schema-coverage.test.ts]: I believe this will require exposing schema of tools from MCP? Right?

File: package.json [hunk 4]
```diff
@@ -12,7 +12,8 @@
     "db:generate": "drizzle-kit generate",
     "db:migrate": "drizzle-kit migrate",
     "db:push": "drizzle-kit push",
-    "db:studio": "drizzle-kit studio"
+    "db:studio": "drizzle-kit studio",
+    "test:mcp": "vitest run test/mcp-schema-coverage"
```
Comment (mujahidkay) [inline package.json]:
```suggestion
    "test": "vitest",
    "test:run": "vitest run"
```

File: test/mcp-schema-coverage.test.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
+/**
+ * MCP Tool Schema Coverage Test
+ *
+ * This regression test verifies that all tools exposed by MCP servers
+ * have corresponding schema definitions in the /lib/mcp/ directory.
+ *
+ * Purpose:
+ * - Ensure schema definitions stay synchronized with MCP server implementations
+ * - Catch missing schemas when MCP servers add/modify tools
```
Comment (mujahidkay) [inline test/mcp-schema-coverage.test.ts]: This is incorrect. We currently aren't (and can't) catch tool schema modification

File: test/mcp-schema-coverage.test.ts [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+import { agoricMcpDevopsToolSchemas } from '@/lib/mcp/agoric-devops-tool-schemas';
+import { ymaxMcptoolSchemas } from '@/lib/mcp/ymax-tool-schemas';
+import { fetchMcpServerTools, cleanupMcpClients, type ServerConfig, type CachedClient } from './test-helpers';
+
+const MCP_SERVERS: Record<string, ServerConfig> = {
+    agoric: {
+        name: 'Agoric MCP Server',
+        url: 'https://agoric-mcp-server.agoric-core.workers.dev/sse',
+        schemas: agoricMcpToolSchemas,
+        schemaFile: 'lib/mcp/agoric-tool-schemas.ts',
```
Comment (mujahidkay) [inline test/mcp-schema-coverage.test.ts]: this field is unneeded, only used in logs, which themselves are unnecessary

File: test/mcp-schema-coverage.test.ts (hunk 2 above)
Comment (waqasshaukat) [reply test/mcp-schema-coverage.test.ts]: @Muneeb147 Good catch! Actually, there were multiple tests before cleaning up the PR. The mcpClientCache was serving its purpose then. Now it's redundant one. Removing it in the next commit.

File: test/mcp-schema-coverage.test.ts (hunk 3 above)
Comment (waqasshaukat) [reply test/mcp-schema-coverage.test.ts]: @Muneeb147 Due to version difference of zod in MCP servers and in chat, there are some limitations of matching the schema. Due to this, the scope of this PR has been limited to presence of schemas of relevant tools. @mujahidkay is working on updating the zod version in MCP servers. After that we were be able to match them as well.

Comment (mujahidkay) [inline test/mcp-schema-coverage.test.ts]: Per our discussion today, did you explore how to write parameterized tests in `vitest`? Please go through https://vitest.dev/api/#test-each or https://vitest.dev/api/#test-for and make the appropriate adjustments

File: test/mcp-schema-coverage.test.ts [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+    Object.entries(MCP_SERVERS).forEach(([serverKey, serverConfig]) => {
+        describe(serverConfig.name, () => {
+            it('should have matching schemas for all MCP server tools', { timeout: TEST_TIMEOUT }, async () => {
+                const serverTools = await fetchMcpServerTools(serverKey, serverConfig, mcpClientCache);
+                const definedSchemas = Object.keys(serverConfig.schemas);
+
+                console.log(`\n ${serverConfig.name} Validation:`);
+                console.log(`   Server tools: ${serverTools.length}`);
+                console.log(`   Defined schemas: ${definedSchemas.length}`);
```
Comment (mujahidkay) [inline test/mcp-schema-coverage.test.ts]: There's rarely any need for logs in tests... Please remove.

File: test/mcp-schema-coverage.test.ts [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+                expect(
+                    definedSchemas.length,
+                    `Schema count mismatch in ${serverConfig.schemaFile}: expected ${serverTools.length} but got ${definedSchemas.length}`
+                ).toBe(serverTools.length);
```
Comment (mujahidkay) [inline test/mcp-schema-coverage.test.ts]:
```suggestion
                expect(
                    definedSchemas.length
                ).toBe(serverTools.length);
```
i believe this statement alone is very aptly descriptive, there's no need for extra text.

File: test/mcp-schema-coverage.test.ts [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+                const missingSchemas = serverTools.filter(
+                    (tool) => !definedSchemas.includes(tool)
+                );
+
+                if (missingSchemas.length > 0) {
+                    console.error('\n Missing schemas for tools:');
+                    missingSchemas.forEach((tool) => console.error(`   - ${tool}`));
+                }
```
Comment (mujahidkay) [inline test/mcp-schema-coverage.test.ts]:
```suggestion
```

File: test/mcp-schema-coverage.test.ts [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+                expect(
+                    missingSchemas,
+                    `Missing schemas in ${serverConfig.schemaFile} for tools: ${missingSchemas.join(', ')}`
+                ).toHaveLength(0);
```
Comment (mujahidkay) [inline test/mcp-schema-coverage.test.ts]:
```suggestion
                expect(
                    missingSchemas
                ).toHaveLength(0);
```

File: test/mcp-schema-coverage.test.ts [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+            });
+        });
+    });
+});
```
Comment (mujahidkay) [inline test/mcp-schema-coverage.test.ts]:
```suggestion
});

```

**Review by mujahidkay — CHANGES_REQUESTED** [review]
I don't see the point of `missingSchemas` part of the test. A count mismatch already tells us that. A better way would be to do snapshot matching of tool names with defined schemas in this repo, and those in remote MCP servers.

File: test/mcp-schema-coverage.test.ts (hunk 3 above)
Comment (mujahidkay) [reply test/mcp-schema-coverage.test.ts]: > Due to version difference of zod in MCP servers and in chat

How did you come to this conclusion? This is definitely not the main reason we can't do schema matching.

File: test/mcp-schema-coverage.test.ts (hunk 10 above)
Comment (waqasshaukat) [reply test/mcp-schema-coverage.test.ts]: > 

I think the original line would help user/tester to find quickly which tools schemas are missing.

File: test/mcp-schema-coverage.test.ts (hunk 3 above)
Comment (waqasshaukat) [reply test/mcp-schema-coverage.test.ts]: > > Due to version difference of zod in MCP servers and in chat
> 
> How did you come to this conclusion? This is definitely not the main reason we can't do schema matching.

So far I know, we can compare them by converting zode schema to json schema, for that we have to use a dependency "zod-to-json-schema" then deep comparison. This approach might has some edge cases.

**Comment by waqasshaukat** [comment]
> I don't see the point of `missingSchemas` part of the test. A count mismatch already tells us that. A better way would be to do snapshot matching of tool names with defined schemas in this repo, and those in remote MCP servers.

`
           const missingSchemas = serverTools.filter(
                (tool) => !definedSchemas.includes(tool)
            );

            expect(
                missingSchemas,
                `Missing schemas in ${serverConfig.name} for tools: ${missingSchemas.join(', ')}`
            ).toHaveLength(0);
`
Above code is checking for missing schemas - tools that exist on the MCP server but don't have corresponding schema definitions in our codebase. Count mismatch will not guaranteed that agoric-mcp-chat has all the schemas e.g. if an old tool is replaced by a new one(for some reason) then the count will remains same.

**Comment by mujahidkay** [comment]
> > I don't see the point of `missingSchemas` part of the test. A count mismatch already tells us that. A better way would be to do snapshot matching of tool names with defined schemas in this repo, and those in remote MCP servers.
> 
> ` const missingSchemas = serverTools.filter( (tool) => !definedSchemas.includes(tool) );
> 
> ```
>         expect(
>             missingSchemas,
>             `Missing schemas in ${serverConfig.name} for tools: ${missingSchemas.join(', ')}`
>         ).toHaveLength(0);
> ```
> 
> ` Above code is checking for missing schemas - tools that exist on the MCP server but don't have corresponding schema definitions in our codebase. Count mismatch will not guaranteed that agoric-mcp-chat has all the schemas e.g. if an old tool is replaced by a new one(for some reason) then the count will remains same.

Precisely what a snapshot would help with. 
1. asserts num tools match
2. assert definedSchemas to match snapshot (will fail when someone adds new tool to definedSchemas [requires snapshot update], and the test outputs the difference)
3. assert toolSchemas to match snapshot (will fail when a new tool is added to mcp server [requires snapshot update which will be done when porting the schema to this repo], and the test outputs the difference

**Review by mujahidkay — APPROVED** [review]
Thanks for addressing comments. Please clean up the commits or squash while merging.
